### ISSUE TITLE
Leaving duplicate workspace nodes When refreshing

### ISSUE THREAD

**Issue by yiliang114** [microsoft/vscode#171458]
Reproduction step:
1. Open `https://vscode.dev/`, then open any project from github, such as `https://vscode.dev/github/yiliang114/vscode`
2. After the file tree is loaded, refresh the page directly without any other operations. 
3. You can see that the data of the file tree has disappeared, leaving only a temporary folder named workspace name, but in fact, there is already a workspace name on tree item.

![bug15](https://user-images.githubusercontent.com/11473889/212802254-b426cd6a-190e-45ad-b665-29e706b0930d.gif)

![image](https://user-images.githubusercontent.com/11473889/212802301-ff1cc655-f3f3-4ba7-a8a1-524d5d5360f8.png)

Version: 1.74.3
Commit: 97dec172d3256f8ca4bfb2143f3f76b503ca0534
User Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36
Embedder: vscode.dev

**Comment by lramos15** [comment]
@jrieken I think this is FS provider related. Let me know if you disagree, it also might be a lifecycle thing cc @bpasero

**Comment by jrieken** [comment]
I don't think so, I remember that the explorer always had issues with that. Since it's the explorer showing the error I suggest that the investigation starts there. It's likely that it gives us to restore a persisted state before the actual extension registers the FS

**Comment by bpasero** [comment]
cc @sandy081, I remember some tricks with how extension activation works with custom schemes, but I forgot.

**Comment by sandy081** [comment]
In configuration land I wait until the FSP is registered like this

https://github.com/microsoft/vscode/blob/1ed7421b6d1edba54e51c7a8fb27201687cd2470/src/vs/workbench/services/configuration/browser/configuration.ts#L667

**Comment by joyceerhl** [comment]
This scenario happens when the workbench has already loaded and the page is being refreshed.

1. The page reload triggers `stopExtensionHosts` in `abstractExtensionHost.ts`
2. In this case the FS provider is coming from an extension, so stopping the extension host also disposes the means the FS provider gets disposed
3. The explorer service listens for `onDidChangeFileSystemProviderRegistrations` and resets the tree input https://github.com/microsoft/vscode/blob/8cd2eb1b3273e8f874c62e0b508ced1d3f34422a/src/vs/workbench/contrib/files/browser/explorerService.ts#L122
4. This leads to the explorer tree view getting rerendered
5. The explorer has a decorations provider which runs after the rerender and it calls `stat` on the resource. At this point there is no filesystem capable of handling the stat anymore, so you get the `! Unable to resolve workspace folder` decoration, just before the page reloads

This seems like mostly a visual problem rather than missing/broken functionality, maybe the explorer view can avoid setting the tree input and triggering the layout if the workbench is about to go down anyway?

**Comment by yiliang114** [comment]
> This scenario happens when the workbench has already loaded and the page is being refreshed.
> 
> 1. The page reload triggers `stopExtensionHosts` in `abstractExtensionHost.ts`
> 2. In this case the FS provider is coming from an extension, so stopping the extension host also disposes the means the FS provider gets disposed
> 3. The explorer service listens for `onDidChangeFileSystemProviderRegistrations` and resets the tree input https://github.com/microsoft/vscode/blob/8cd2eb1b3273e8f874c62e0b508ced1d3f34422a/src/vs/workbench/contrib/files/browser/explorerService.ts#L122
> 4. This leads to the explorer tree view getting rerendered
> 5. The explorer has a decorations provider which runs after the rerender and it calls `stat` on the resource. At this point there is no filesystem capable of handling the stat anymore, so you get the `! Unable to resolve workspace folder` decoration, just before the page reloads
> 
> This seems like mostly a visual problem rather than missing/broken functionality, maybe the explorer view can avoid setting the tree input and triggering the layout if the workbench is about to go down anyway?

Before refreshing the page, is the `setTreeInput ` called in  onDidChangeFileSystemProviderRegistrations callback to reduce memory leaks?  If not, I think the data of the file tree is not reset, will look better to experience it when the `onDidChangeFileSystemProviderRegistrations` triggered by refresh. It is a bit strange that the data disappears faster than refresh behavior.

**Comment by yiliang114** [comment]
`onDidChangeFileSystemProviderRegistrations ` Triggered by events here
https://github.com/microsoft/vscode/blob/25514d899f4bd877803defb213719f8af3d997d3/src/vs/platform/files/common/fileService.ts#L73
